[TexMap] Damage rectangles are too big when scrolling

Scony opened the pull request:
#### cbe664e3befcd7e3db8ffb0350e60150a2deecd5
<pre>
[TexMap] Incorrect damage when 3d transform added dynamically
<a href="https://bugs.webkit.org/show_bug.cgi?id=288471">https://bugs.webkit.org/show_bug.cgi?id=288471</a>

Reviewed by NOBODY (OOPS!).

This change fixes the damage inference logic in a way that newly added
TextureMapperLayers are not taking the damage from before tranfrom
change into account.

* LayoutTests/platform/glib/TestExpectations:
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::collectDamageSelf):
(WebCore::TextureMapperLayer::damageWholeLayerDueToTransformChange):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
</pre>

spenap commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -438,13 +477,14 @@ void TextureMapperLayer::collectDamageSelf(TextureMapperPaintOptions& options, D
     // Inferred damage always takes precedence over any other kind of damage as it stores
     // the damage that is a result of the layer-level operations.
     if (!m_inferredGlobalDamage.isEmpty() || !m_inferredLayerDamage.isEmpty()) {
+        auto clipBounds = options.textureMapper.clipBounds();
```
I was checking how expensive it was to copy what `clipBounds()` returned (it's cheap, it's an `IntRect`, although you could also do `const auto& clipBounds` to bind to it). 

Then I thought `why is `clipBounds()` not `const`? I looked at the implementation, and it does

```
 clipStack().current().scissorBox;
```

neither `clipStack()` nor `current()` are `const`. It makes sense to have `non-const` options as often they will be used to modify things. But I wonder if we should have const overloads, too, as some of the usages are read-only.

But anyway. Beyond the `const auto&` suggestion, the rest is outside of the scope of this PR

Scony replied:
Noted down, thanks!

Scony committed:
[TexMap] Damage rectangles are too big when scrolling
https://bugs.webkit.org/show_bug.cgi?id=288468

Reviewed by Carlos Garcia Campos.

This change fixes the damage rectangles generated on scrolling
by maintaining a clipping rect and using it to clip those
based on the layer ancestors.

Also, this change fixes a few layout tests that were not correct
in the sense that they could be affected by scrollbars
depending on the platform the tests were run on.

* LayoutTests/platform/glib/TestExpectations:
* LayoutTests/platform/glib/damage/accelerated-canvas-2d-fillRect.html:
* LayoutTests/platform/glib/damage/scrolling-container-001.html:
* LayoutTests/platform/glib/damage/scrolling-container-003.html:
* LayoutTests/platform/glib/damage/scrolling-container-004.html:
* LayoutTests/platform/wpe/TestExpectations:
* Source/WebCore/platform/graphics/texmap/TextureMapper.cpp:
(WebCore::TextureMapper::beginClipWithoutApplying):
(WebCore::TextureMapper::endClipWithoutApplying):
* Source/WebCore/platform/graphics/texmap/TextureMapper.h:
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::collectDamageSelfAndChildren):
(WebCore::TextureMapperLayer::collectDamageSelf):

Canonical link: https://commits.webkit.org/291382@main
